tools/sva/src/parser_test.js

```diff
@@ -0,0 +1,458 @@
+
+        let inst = ast.instruction(0);
+        assert.equal("OpImageWrite", inst.name());
+        assert.lengthOf(inst.operands(), 4);
+        assert.equal("ConstOffset", inst.operand(3).name());
+      });
+    });
+
+    describe("*", () => {
+      it("skips if missing", () => {
+        let input = `OpEntryPoint Fragment %main "main"
+OpKill`;
+
+        let l = new Lexer(input);
+        let p = new Parser(grammar, l);
+
+        let ast = p.parse();
+        assert.exists(ast);
+        assert.lengthOf(ast.instructions(), 2);
+
+        let inst = ast.instruction(0);
+        assert.equal("OpEntryPoint", inst.name());
+        assert.lengthOf(inst.operands(), 3);
+        assert.equal("main", inst.operand(2).name());
+      });
+
+      it("extracts one if available", () => {
+        let input = `OpEntryPoint Fragment %main "main" %2
+OpKill`;
+
+        let l = new Lexer(input);
+        let p = new Parser(grammar, l);
+
+        let ast = p.parse();
+        assert.exists(ast);
+        assert.lengthOf(ast.instructions(), 2);
+
+        let inst = ast.instruction(0);
+        assert.equal("OpEntryPoint", inst.name());
+        assert.lengthOf(inst.operands(), 4);
+        assert.equal("2", inst.operand(3).name());
+      });
+
+      it("extracts multiple if available", () => {
+        let input = `OpEntryPoint Fragment %main "main" %2 %3 %4 %5
+OpKill`;
+
+        let l = new Lexer(input);
+        let p = new Parser(grammar, l);
+
+        let ast = p.parse();
+        assert.exists(ast);
+        assert.lengthOf(ast.instructions(), 2);
+
+        let inst = ast.instruction(0);
+        assert.equal("OpEntryPoint", inst.name());
+        assert.lengthOf(inst.operands(), 7);
+        assert.equal("2", inst.operand(3).name());
+        assert.equal("3", inst.operand(4).name());
+        assert.equal("4", inst.operand(5).name());
+        assert.equal("5", inst.operand(6).name());
+      });
+    });
+  });
+
+  describe("extended instructions", () => {
+    it("errors on non-glsl extensions", () => {
+      let input = "%1 = OpExtInstImport \"OpenCL.std.100\"";
+
+      let l = new Lexer(input);
+      let p = new Parser(grammar, l);
+
+      assert.isUndefined(p.parse());
+    });
+
+    it("handles extended instructions", () => {
+      let input = `%1 = OpExtInstImport "GLSL.std.450"
+  %44 = OpExtInst %7 %1 Sqrt %43`;
+
+      let l = new Lexer(input);
+      let p = new Parser(grammar, l);
+
+      let ast = p.parse();
+      assert.exists(ast, p.error);
+      assert.lengthOf(ast.instructions(), 2);
+
+      let inst = ast.instruction(1);
+      assert.lengthOf(inst.operands(), 5);
+      assert.equal(31, inst.operand(3).value());
+      assert.equal("Sqrt", inst.operand(3).name());
+      assert.equal(43, inst.operand(4).value());
+      assert.equal("43", inst.operand(4).name());
+    });
+  });
+
+  it.skip("handles spec constant ops", () => {
+    // let input = "%sum = OpSpecConstantOp %i32 IAdd %a %b";
+  });
```

Comment:
Please add this stress test from OpCopyMemory. It uses two optional memory access operands, each of which has its own mixed set of enums, some of which with its own parameters. 

https://github.com/KhronosGroup/SPIRV-Tools/blob/master/test/text_to_binary.memory_test.cpp#L274

Here's the whole test, showing you exactly which words are produced.  You can figure out how the mapping goes by staring at it...

```
  std::string spirv =
      "OpCopyMemory %1 %2 Volatile|Nontemporal|"
      "MakePointerVisible %3 "
      "Aligned|MakePointerAvailable|NonPrivatePointer 16 %4\n";
  EXPECT_THAT(CompiledInstructions(spirv),
              Eq(MakeInstruction(SpvOpCopyMemory, {1, 2, 21, 3, 42, 16, 4})));
  std::string disassembly =
      EncodeAndDecodeSuccessfully(spirv, SPV_BINARY_TO_TEXT_OPTION_NONE);
  EXPECT_THAT(disassembly, Eq(spirv));
```

Reply:
Ugh, gross. That's going to fail, heh.

Reply:
Added a skip test, will work out post land.

Reply:
yeah spir-v syntax is deliberately LR parsing. Need to keep a stack of expected tokens, and reading the mask makes you push stack of more expected tokens.

Reply:
See https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/text.cpp#L559

Reply:
ack.